File: 2021/09/10/ZigZag.java
/**
 * https://leetcode.com/problems/zigzag-conversion/
 * @author Andrew Jarombek
 * @since 9/10/2021
 */


import java.util.ArrayList;
import java.util.List;

public class ZigZag {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        int currentRow = 0;
        boolean movingDown = true;

        List<StringBuilder> sbl = new ArrayList<>();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < numRows; i++) {
            sbl.add(new StringBuilder());
        }

        for (int i = 0; i < s.length(); i++) {
            sbl.get(currentRow).append(s.charAt(i));

            if (movingDown) {
                if (currentRow == numRows - 1) {
                    movingDown = false;
                    currentRow--;
                } else {
                    currentRow++;
                }
            } else {
                if (currentRow == 0) {
                    movingDown = true;
                    currentRow++;
                } else {
                    currentRow--;
                }
            }
        }

        for (StringBuilder sbi: sbl) {
            sb.append(sbi);
        }

        return sb.toString();
    }
}

File: 2022/01/14/zig_zag.py
"""
https://leetcode.com/problems/zigzag-conversion/
Author: Andrew Jarombek
Date: 1/14/2022
"""


def convert(s: str, num_rows: int) -> str:
    if num_rows == 1:
        return s

    rows = [[] for row in range(num_rows)]
    i = 0
    up = False

    for c in s:
        rows[i].append(c)

        if up:
            if i == 0:
                up = False
                i += 1
            else:
                i -= 1
        else:
            if i == num_rows - 1:
                up = True
                i -= 1
            else:
                i += 1

    return "".join("".join(row) for row in rows)
